Include/internal/pycore_lock.h

```diff
@@ -0,0 +1,161 @@
+// Lightweight locks and other synchronization mechanisms.
+//
+// These implementations are based on WebKit's WTF::Lock. See
+// https://webkit.org/blog/6161/locking-in-webkit/ for a description of the
+// design.
+#ifndef Py_INTERNAL_LOCK_H
+#define Py_INTERNAL_LOCK_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include "pycore_time.h"          // _PyTime_t
+
+
+// A mutex that occupies one byte. The lock can be zero initialized.
+//
+// Only the two least significant bits are used. The remaining bits should be
+// zero:
+// 0b00: unlocked
+// 0b01: locked
+// 0b10: unlocked and has parked threads
+// 0b11: locked and has parked threads
+//
+// Typical initialization:
+//   PyMutex m;
+//   memset(&m, 0, sizeof(m));
+//
+// Typical usage:
+//   PyMutex_Lock(&m);
+//   ...
+//   PyMutex_Unlock(&m);
+typedef struct _PyMutex {
+    uint8_t v;
+} PyMutex;
+
+typedef enum {
+    _Py_UNLOCKED = 0,
+    _Py_LOCKED = 1,
+    _Py_HAS_PARKED = 2,
+} _PyMutex_State;
+
+// (private) slow path for locking the mutex
+PyAPI_FUNC(void) _PyMutex_LockSlow(PyMutex *m);
+
+// (private) slow path for unlocking the mutex
+PyAPI_FUNC(void) _PyMutex_UnlockSlow(PyMutex *m);
+
+// Locks the mutex.
+//
+// If the mutex is currently locked, the calling thread will be parked until
+// the mutex is unlocked. If the current thread holds the GIL, then the GIL
+// will be released while the thread is parked.
+static inline void
+PyMutex_Lock(PyMutex *m)
+{
+    uint8_t expected = _Py_UNLOCKED;
```

Comment:
Would it make sense to make this const?

```suggestion
    const uint8_t expected = _Py_UNLOCKED;
```

What about static?

FWIW, it isn't obvious to me why we have to pass a pointer for `expected`.

Reply:
It can't be const. The compare_exchange operations take a pointer to "expected". If the comparison fails, `expected` is updated with the current value. That's not relevant here, but can be useful in compare_exchange loops.

In the first version of the pyatomic.h PR, I had compare_exchange operations take a value (not a pointer) for expected, because I thought it was easier to understand, but the feedback from reviewers in #108338 was that it was more important to match the C11 API.

Reply:
Thanks for clarifying.  That makes sense.